hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/ListSubcommand.java

```diff
@@ -105,12 +115,49 @@ public void execute(ScmClient scmClient) throws IOException {
           ReplicationType.fromProto(type),
           replication, new OzoneConfiguration());
     }
-    List<ContainerInfo> containerList =
-        scmClient.listContainer(startId, count, state, type, repConfig);
 
-    // Output data list
-    for (ContainerInfo container : containerList) {
-      outputContainerInfo(container);
+    int maxCountAllowed = parent.getParent().getOzoneConf()
+        .getInt(ScmConfigKeys.HDDS_CONTAINER_LIST_MAX_COUNT,
+            ScmConfigKeys.HDDS_CONTAINER_LIST_MAX_COUNT_DEFAULT);
+
+    ContainerListResult containerListAndTotalCount;
+
+    if (!all) {
+      if (count > maxCountAllowed) {
+        System.err.printf("Attempting to list the first %d records of containers." +
+            " However it exceeds the cluster's current limit of %d. The results will be capped at the" +
+            " maximum allowed count.%n", count, ScmConfigKeys.HDDS_CONTAINER_LIST_MAX_COUNT_DEFAULT);
+        count = maxCountAllowed;
+      }
+      containerListAndTotalCount = scmClient.listContainer(startId, count, state, type, repConfig);
+      for (ContainerInfo container : containerListAndTotalCount.getContainerInfoList()) {
+        outputContainerInfo(container);
+      }
+
+      if (containerListAndTotalCount.getTotalCount() > count) {
+        System.err.printf("Container list is truncated as it is too long. " +
+                        "Displayed the first %d records of %d total containers.%n",
+                count, containerListAndTotalCount.getTotalCount());
+      }
+    } else {
+      // Batch size is either count passed through cli or maxCountAllowed
+      int batchSize = (count > 0) ? count : maxCountAllowed;
+      long currentStartId = startId;
+      int fetchedCount;
+
+      do {
+        // Fetch containers in batches of 'batchSize'
+        containerListAndTotalCount = scmClient.listContainer(currentStartId, batchSize, state, type, repConfig);
+        fetchedCount = containerListAndTotalCount.getContainerInfoList().size();
+
+        for (ContainerInfo container : containerListAndTotalCount.getContainerInfoList()) {
+          outputContainerInfo(container);
+        }
+
+        if (fetchedCount > 0) {
+          currentStartId = containerListAndTotalCount.getContainerInfoList().get(fetchedCount - 1).getContainerID();
+        }
+      } while (fetchedCount == batchSize);
```

Comment:
If client and server configuration for MAX is different, in this case, it will not give all count. Can change this while loop, fetchedCount > 0, so that it will have extra iteration to confirm if no records.